src/Mvc/Mvc.ApiExplorer/src/EndpointMetadataApiDescriptionProvider.cs

```diff
@@ -243,15 +243,23 @@ private ApiDescription CreateApiDescription(RouteEndpoint routeEndpoint, string
                 responseType = typeof(void);
             }
 
-            var responseMetadata = endpointMetadata.GetOrderedMetadata<IApiResponseMetadataProvider>();
+            // For back-compat, we support attributes (which implement the IApiResponseMetadataProvider) interface
+            // and types added via the extension methods (which implement IProducesResponseTypeMetadata).
+            var responseProviderMetadata = endpointMetadata.GetOrderedMetadata<IApiResponseMetadataProvider>();
+            var producesResponseMetadata = endpointMetadata.GetOrderedMetadata<IProducesResponseTypeMetadata>();
             var errorMetadata = endpointMetadata.GetMetadata<ProducesErrorResponseTypeAttribute>();
             var defaultErrorType = errorMetadata?.Type ?? typeof(void);
             var contentTypes = new MediaTypeCollection();
 
-            var responseMetadataTypes = ApiResponseTypeProvider.ReadResponseMetadata(
-                responseMetadata, responseType, defaultErrorType, contentTypes);
+            var responseProviderMetadataTypes = ApiResponseTypeProvider.ReadResponseMetadata(
+                responseProviderMetadata, responseType, defaultErrorType, contentTypes);
+            var producesResponseMetadataTypes = ReadResponseMetadata(producesResponseMetadata, responseType);
 
-            if (responseMetadataTypes.Count > 0)
+            // We favor types added via the extension methods (which implements IProducesResponseTypeMetadata)
+            // over those that are added via attributes.
+            var responseMetadataTypes = producesResponseMetadataTypes.Concat(responseProviderMetadataTypes);
+
+            if (responseMetadataTypes.Any())
```

Comment:
Remove?
```suggestion
            if (responseMetadataTypes.Any())
```

Reply:
We still need to handle the case where the user doesn't annotate their endpoint in any way (via attribute or metadata).